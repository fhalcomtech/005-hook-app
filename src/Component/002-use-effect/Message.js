import React, {useEffect} from "react";

export const Message = ({text}) => {
    useEffect(() => {
        return () => {
            console.log('good level');
        };
    }, []);

    return (<>
        <h1>Que mas {text}</h1>
    </>);
}