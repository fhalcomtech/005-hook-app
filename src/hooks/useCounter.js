import { useState } from "react";

export const useCounter = ({initial = 0, increment=1}) => {
 const [counter, setCounter] = useState(initial);
 const more = () => (setCounter(counter+increment));
 const less = () => (setCounter(counter-increment));
 const reset = () => (setCounter(initial));
 return ({counter,more,less,reset});
}
