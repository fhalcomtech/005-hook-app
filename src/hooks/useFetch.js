import {useEffect, useState} from "react";

export const useFetch = (url) =>
{
    const [data, setData] = useState({data: null, loading: true, error: false});
    useEffect(() => {
        setData({data: null, loading: true, error: null})
        fetch(url)
            .then(data => data.json())
            .then(json => {setData({data: json, loading: false, error: false});})
            .catch(err=>{setData({data: null, loading: false, error: true});})
    }, [url]);
    return data;
}