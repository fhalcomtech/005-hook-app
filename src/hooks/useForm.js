import React, { useState } from 'react'

export const useForm = (initialData = {}) => {

    const [formData, setFormData] = useState(initialData);

    const handlerInputChange = ({target}) => {setFormData({...formData, [target.name]: target.value})}

    return [formData, handlerInputChange];
}
