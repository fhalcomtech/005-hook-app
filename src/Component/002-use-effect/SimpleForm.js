import React, { useEffect, useState } from 'react'
import {Message} from "./Message";

export const SimpleForm = () => {
    const [formData, setFormData] = useState({user: '', email:''});
    const {user,email} = formData;
    const handlerChange = (e) => {
        setFormData({...formData, [e.target.name]: e.target.value})
    }

    useEffect(()=>{console.log('Inicio el componente')},[])

    useEffect(()=>{
        setFormData({...formData, email: user.length})
    },[user]);

    return (
        <>
            <h1>SimpleForm Login</h1>
            <div className = "container">
            <input 
            type="text"
            name="user"
            placeholder='write here'
            className="form-control"
            value={user}
            onChange={handlerChange}
            />

<input
    type="email"
    name="email"
    placeholder=''
    value={email}
    className="form-control mt-2"
    onChange={handlerChange}
    />
                <div>
                    {user === 'fhalcom' && <Message text={user}/>}
                </div>
            </div>
        </>
  )
}
