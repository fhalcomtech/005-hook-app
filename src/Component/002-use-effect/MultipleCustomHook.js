import React from 'react'
import {useCounter} from "../../hooks/useCounter";
import {useFetch} from "../../hooks/useFetch";

export const MultipleCustomHook = () => {
    const {counter, less, more, reset} = useCounter({initial: 1, increment: 1});
    const {data, loading, error} = useFetch(`https://www.breakingbadapi.com/api/quotes/${(counter<1 || counter > 10)?reset():counter}`);
    const {author, quote, series} = !loading && data[0];
  return (
      <>
          <div className="container">
              <h1><span className="badge bg-primary">MultipleCustomHook</span></h1>
          </div>
          {
              loading ? (<div className="container alert alert-primary">Loading...</div>):( <div className="container card-container">
                  <div className="card">
                      <div className="card-body">
                          <h5 className="card-title"><span className="badge bg-secondary">{author}</span></h5>
                          <h6 className="card-subtitle mb-2 text-muted"><span className="badge bg-secondary">{series}</span></h6>
                          <p className="card-text">{quote}</p>
                      </div>
                  </div>
                  <div className="container pt-4">
                    <button className="btn btn-danger" name="btn-prev" onClick={()=>{less()}}>prev</button>
                    <button className="btn btn-primary"  name="btn-next" onClick={()=>{more()}}>next</button>
                  </div>
              </div>)
          }
      </>
  )
}