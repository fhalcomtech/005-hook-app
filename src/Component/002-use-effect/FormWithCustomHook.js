import React, { useEffect, useState } from "react";
import { useForm } from "../../hooks/useForm";

export const FormWithCustomHook = () => {
  const [formData, handlerChange] = useForm({
    user: "",
    email: "",
    password: "",
  });
  const { user, email, password } = formData;

  useEffect(() => {
    console.log("Inicio el componente");
  }, []);


  return (
    <>
      <h1>SimpleForm Login</h1>
      <h3>Su nombre tiene {user.length} letras</h3>
      <div className="container">
        <input
          type="text"
          name="user"
          placeholder="write here"
          className="form-control"
          value={user}
          onChange={handlerChange}
        />

        <input
          type="email"
          name="email"
          placeholder=""
          value={email}
          className="form-control mt-2"
          onChange={handlerChange}
        />

        <input
          type="password"
          name="password"
          placeholder="*******"
          value={password}
          className="form-control mt-2"
          onChange={handlerChange}
        />
      </div>
    </>
  );
};
