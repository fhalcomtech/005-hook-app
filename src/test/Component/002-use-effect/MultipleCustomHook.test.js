import React from "react";
import {shallow} from "enzyme";
import {MultipleCustomHook} from "../../../Component/002-use-effect/MultipleCustomHook";
import {useFetch} from "../../../hooks/useFetch";
import {useCounter} from "../../../hooks/useCounter";

jest.mock("../../../hooks/useFetch");
jest.mock("../../../hooks/useCounter");

describe("Testing MultipleCustomHooks file", ()=>{
    useCounter.mockReturnValue({counter:1,more:()=>{},less:()=>{},reset:()=>{}})

    test("MultipleCustomHooks create new snapshopt",()=>{
         useFetch.mockReturnValue({data: null, loading: true, error: false});
         const wrapper = shallow(<MultipleCustomHook/>);
         expect(wrapper).toMatchSnapshot();
    });

    test("Testing useFetch from MultipleCustomHooks", (done)=>{
        const mockData = {author: "Walter", series: "Fhalcom", quote:"No tiene sentido"};
        useFetch.mockReturnValue({data: [mockData], loading: false, error: false})
        const wrapper = shallow(<MultipleCustomHook/>);
        const cartTitle = wrapper.find(".card .card-body .card-title .bg-secondary").text().trim();
        const cartSubtitle = wrapper.find(".card .card-body .card-subtitle .bg-secondary").text().trim();
        const cartText = wrapper.find(".card .card-body .card-text").text().trim();
        expect(cartTitle).toEqual(mockData.author);
        expect(cartSubtitle).toEqual(mockData.series);
        expect(cartText).toEqual(mockData.quote);
        done();
    });
    test("Testing useCounter from MultipleCustomHooks", (done)=>{
        const wrapper = shallow(<MultipleCustomHook/>);
        done();
    });
})
