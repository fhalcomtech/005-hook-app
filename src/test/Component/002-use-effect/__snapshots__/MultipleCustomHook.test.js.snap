// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing MultipleCustomHooks file MultipleCustomHooks create new snapshopt 1`] = `
<Fragment>
  <div
    className="container"
  >
    <h1>
      <span
        className="badge bg-primary"
      >
        MultipleCustomHook
      </span>
    </h1>
  </div>
  <div
    className="container alert alert-primary"
  >
    Loading...
  </div>
</Fragment>
`;
